### ISSUE TITLE
Authentication issue on tenant identifier changes

### ISSUE THREAD

**Issue by EngrKhizarIqbal** [Finbuckle/Finbuckle.MultiTenant#307]
Hi, first thanks for this awesome library. Lots of love.

I've been working on a SaaS project and everything works fine and as expected. However, I've been asked to add a feature where users can modify tenant identifiers (we are using tenant identifiers as subdomains). I've implemented this feature but trapped into an issue with the authentication. When a user updates their tenant identifier (subdomain), the system no longer able to authenticate them.

We are logging the user actions and I can see in the logs the login method runs successfully and `_signInManager.PasswordSignInAsync(...)` returns success, however in the database logs table (we are using Serilog), I see the following errors.

1. AuthenticationScheme: "Cookies" was challenged.
3. Authorization failed.
3. ``` {
   "AuthenticationScheme": "Cookies",
   "EventId": {
     "Id": 12,
     "Name": "AuthenticationSchemeChallenged"
   },
   "SourceContext": "Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler",
   "RequestId": "80000043-0000-cb00-b63f-84710c7967bb",
   "RequestPath": "/Account/Logout",
   "SpanId": "983a89a9c9231744",
   "TraceId": "31d959065f6fc240b06405e450f95ba7",
   "ParentId": "0000000000000000",
   "Application": "Quody"
   }
   ```

I've tried some solutions but none works. I've made a sample project that replicates this issue. You can clone it from here [finbuckle-auth-issue-sample](https://github.com/EngrKhizarIqbal/finbuckle-auth-issue-sample).

A readme file is also included with the required PMC commands to quickly set up databases and SeedData.sql file with the required initial data.

After creating and seeding databases, run the project. log in to the system with the passwod **Admin@1234_** and on the home page, you'll find a form to change the tenant identifier. Update the tenant to whatever value you want and after that, you'll no longer be able to login to the tenant although in DB everything updated and in logs, you can see messages **User with email {0} logged in successfully.**

Any help to solve this issue will be appreciated.

Bests,
Khizar.

**Comment by AndrewTriesToCode** [comment]
Hi, I’m glad you like the library. 

I think the problem is that cookies by default are specified down to the subdomain and when you change the subdomain the browser no longer sends the cookie on subsequent requests. 

I recommend you try to sign out the user and immeditatwly religion then in. There should be a way for you to specify the new subdomain for the new login cookie, either via authentication properties or hooking the cookie events where I think you can override cookie options. 

I’ll take a closer loom tomorrow and let you know if I can get it to work that way. You might beat me to it.

**Comment by EngrKhizarIqbal** [comment]
Thanks for the response. I've already tried to sign out the user on tenant identifier change and redirect users to the new tenant (subdomain) login page but it didn't work. I'm signing out the users on login request but it also didn't work.

I'm on EID holidays from today and will not available for work today and tomorrow. I'll try to check for the cookie events on Wednesday and will update you.

Bests,

**Comment by AndrewTriesToCode** [comment]
To add:

make sure you have a per-tenant options set up for CookieAuthenticationOptions that sets the domain on the cookie to equal the full subdomain + domain for the tenant.

This isn't important normally, but specifically comes into play when you are changing the subdomain for a tenant as part of a request.

When you re-sign in the user we want it to regenerate the cookie options after you have changed the tenant identifier. After determining the desired tenant, you can call `HttpContext.TrySetTenantInfo` which has an option to reset the DI scope -- this will force the cookie options to be regenerated. Then you can call sign-in and the cookies should get set for the new subdomain and if you redirect it should work.

**Comment by EngrKhizarIqbal** [comment]
Hi, I've tried `HttpContext.TrySetTenantInfo` and then immediately sign out the user but it didn't work.
I even tried to clear the cookies on signout in `CookieAuthenticationEvents.Signout` but it didn't work. Seems like the browser is neglecting the cookies on redirect.

I'm working on some other possible solutions and will update you. Have you got a chance to look at the sample project? I've pushed my code.

Bests.
